src/shell/js/composeui-fdc3/src/ComposeUIDesktopAgent.ts

```diff
@@ -0,0 +1,254 @@
+        return new Promise((resolve, reject) => {
+            if(!this.currentChannel){
+                reject(new Error("The current channel have not been set."));
+            } else {
+                resolve(this.currentChannel.broadcast(context));
+            }
+        })
+    }
+
+    //TODO
+    public raiseIntent(intent: string, context: Context, app?: string | AppIdentifier): Promise<IntentResolution> {
+        throw new Error("Not implemented");
+    }
+
+    //TODO
+    public raiseIntentForContext(context: Context, app?: string | AppIdentifier): Promise<IntentResolution> {
+        throw new Error("Not implemented");
+    }
+
+    //TODO
+    public addIntentListener(intent: string, handler: IntentHandler): Promise<Listener> {
+        throw new Error("Not implemented");
+    }
+
+    public addContextListener(contextType?: string | null | ContextHandler, handler?: ContextHandler): Promise<Listener> {
+        return new Promise<ComposeUIListener>(async(resolve, reject) => {
+            if(!this.currentChannel) {
+                reject(new Error("The current channel is null or undefined"));
+                return;
+            }
+            if(typeof contextType == 'function') {
+                reject(new Error("The contextType was type of ContextHandler, which would use a deprecated function, please use string or null for contextType!"));
+                return;
+            }
+            contextType = contextType as string ?? null; 
+            
+            //Resolving the task of subscription to the messageRouter server.
+            const listener = <ComposeUIListener>await this.currentChannel!.addContextListener(contextType, handler!);
+            const context = await this.currentChannel.getCurrentContext(contextType); //TODO: what happens whe a broadcasted message arrives between 2 points
+            await listener.handleContextMessage(context!);
+            this.currentChannelListeners.push(listener);
+            resolve(listener);
+        });
+    }
+
+    public getUserChannels(): Promise<Array<Channel>> {
+        return Promise.resolve(this.userChannels);
+    }
+
+    public joinUserChannel(channelId: string): Promise<void> {
+        return new Promise<void>((resolve, reject) => {
+            let channel = this.userChannels.find(innerChannel => innerChannel.id == channelId);
+            if (!channel) {
+                reject(new Error("Channel couldn't be found in user channels"));
+            } else {
+                this.currentChannel = channel;
+                resolve();
+            }
+        });
+    }
+
+    public getOrCreateChannel(channelId: string): Promise<Channel> {
+        return new Promise<Channel>((resolve, reject) => {
+            let channel = this.userChannels.find(innerChannel => innerChannel.id == channelId);
+            if (!channel) {
+                channel = new ComposeUIChannel(channelId, "app", this.messageRouterClient); //TODO later
```

Comment:
First it looks for the channel id in `userChannels`, but the new channel is added to `appChannels`, is that intentional?

Reply:
OK it has to be an app channel, so the find must be in `appChannels`